examples/blockchain/chain/data.rs

```diff
@@ -0,0 +1,144 @@
+use crate::{chain::BlockNum, network::NetworkData};
+use aleph_bft::OrderedBatch;
+use futures::channel::mpsc::{self, UnboundedReceiver, UnboundedSender};
+use log::debug;
+use parking_lot::Mutex;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+pub(crate) type Data = BlockNum;
+
+pub(crate) struct DataStore {
+    next_message_id: u32,
+    current_block: Arc<Mutex<BlockNum>>,
+    available_blocks: HashSet<BlockNum>,
+    message_requirements: HashMap<u32, usize>,
+    dependent_messages: HashMap<BlockNum, Vec<u32>>,
+    pending_messages: HashMap<u32, NetworkData>,
+    messages_for_member: UnboundedSender<NetworkData>,
+}
+
+impl DataStore {
+    pub(crate) fn new(
+        current_block: Arc<Mutex<BlockNum>>,
+        messages_for_member: UnboundedSender<NetworkData>,
+    ) -> Self {
+        let available_blocks = (0..=*current_block.lock()).collect();
+        DataStore {
+            next_message_id: 0,
+            current_block,
+            available_blocks,
+            message_requirements: HashMap::new(),
+            dependent_messages: HashMap::new(),
+            pending_messages: HashMap::new(),
+            messages_for_member,
+        }
+    }
+
+    fn add_pending_message(&mut self, message: NetworkData, requirements: Vec<BlockNum>) {
+        let message_id = self.next_message_id;
+        self.next_message_id += 1;
+        self.next_message_id %= 1_000_000;
```

Comment:
Hmm, I'm not entirely sure of the consequences of this modulo, is there something bad that can happen if two different messages exist with the same modulo at the same time? 

Alternatively, one can not do modulo, and simply drop old messages, i.e., we keep only the ones in the interval [next_message_id - 10^6, next_message_id], and drop the expired ones from `message_requirements` and `pending_messages` and ignore messages in  `dependent_messages[block]` that have been dropped.

Not sure if this is equivalent to doing modulo? If so, just ignore :)

Reply:
Agh, this is a problem because of the other change you suggested – if we kept requirements as a HashMap then dropping messages would not be a problem. Eh, I'm removing the modulo and this should still work for a loooong time...